eclipse-theia/theia
# Allow opening changes for files associated with custom editors

#### What it does

Allows the user to visually compare two versions of a file that is associated with a custom editor. This is done by opening both versions of the file simultaneously in a side-by-side editor, with each version being opened in a separate custom editor contained in the side-by-side editor.

For example, this feature can be used to help spot the differences between two versions of an image file associated with the custom editor contributed by the VS Code `media-preview` plugin, as illustrated with the following screenshot:

<img width="2032" alt="Screenshot" src="https://github.com/eclipse-theia/theia/assets/8298491/8cdfe884-5718-4bf0-9bd0-f326b07b88e3">

More screenshots illustrating this feature being used with other custom editors:

<img width="1304" alt="Screenshot" src="https://github.com/eclipse-theia/theia/assets/8298491/9bcee523-5275-40e1-8954-ff3b7dab63a9">

<img width="1304" alt="Screenshot" src="https://github.com/eclipse-theia/theia/assets/8298491/8bb529f2-f769-45d0-9ca2-8d5599c52916">

This PR contains a series of individual commits that work together to make it possible for the `CustomEditorOpener` to open a diff-uri in a side-by-side editor containing the corresponding `CustomEditor`s. While the main functionality is contributed by the latest commit, the earlier commits provide various fixes and enhancements required to make it work.

Here is the sequence of commits:

- Fix loading of webview resources that depend on query params
    
    Some resource url (notably git) use the query part of the url to store additional information. The query part of the url was incorrectly being dropped while attempting to load these resources inside of webviews.

- Fix `Error: Unknown Webview` messages in the log

- Fix design and implementation issues surrounding `CustomEditorOpener`
    
    This commit ensures that the promise returned by `CustomEditorOpener.open` will only resolve to a properly initialized and opened `CustomEditorWidget`. In particular, it ensures that the widget is opened according to the specified `WidgetOpenerOptions`, including `widgetOptions.ref` and `mode`.
    
    Essentially, it revises the work done in #9671 and #10580.

- Restore custom editors as part of layout
    
    Fixes an incorrect assumption that a custom editor cannot be restored if no `WebviewPanelSerializer` is registered for its view type. (Actually, custom editors are created and restored using a custom editor provider.)
    
    Also, ensures that `CustomEditorWidget.modelRef` satisfies the shape for the `CustomEditorWidget` defined in `editor.ts` and cannot return `undefined`. (However, `CustomEditorWidget.modelRef.object` can be `undefined` until the custom editor is resolved.)
    
    Fixes #10787

- Fix a race condition when file system provider is activated
    
    When file system provider is activated, wait until it is registered.
    
- git: add support for custom editors
    
    * Uses `OpenerService` instead of `EditorManager` to open editors
    
    * Contributes a `FileSystemProvider` for git-resources
    
    * Fixes an issue with getting blob contents

- custom editor: open a diff-uri in a side-by-side editor
    
    `CustomEditorOpener` is now able to open a diff-uri in a side-by-side editor, which contains the corresponding `CustomEditor`s.
    
    Fixes #9079

#### How to test

There are various configurations that can be used for testing this feature:

* Windows vs. Linux vs. macOS

* Browser vs. Electron

* `vscode.git` vs. (now deprecated) `@theia/git`

This feature can be tested with any custom editor.

A basic test scenario involves making a change to a file associated with a custom editor (e.g., `.jpg`, `.png`, `.pawDraw`, or `.cscratch`) and then opening changes in a side-by-side editor by clicking on the file in the Source Control view.

Since this PR also fixes an issue with custom editors not being restored as part of layout, it is worth checking out that custom editors are now correctly restored, both standalone and as part of side-by-side editors.

Also, it is important to verify that there are no regressions, especially that standalone custom editors can still be opened and work as expected.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
